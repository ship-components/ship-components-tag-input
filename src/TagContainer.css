/**
 *  <SelectBox />
 */
.container {
    width: 100%;
    box-sizing: border-box;
    background-color: #fff;
    border-radius: 2px;
    min-height: 48px;
    display: flex;
    align-items: center;
    padding: 5px 5px 5px 0;
    position: relative;
    -webkit-transition: all 0.2s;
    transition: all 0.2s;

    input[type="text"] {
        color: #333;
        &::-webkit-input-placeholder  { color: #333 }          /* Chrome/Opera/Safari */
        &::-moz-placeholder,          { color: #333 }          /* Firefox 19+ */
        &:-ms-input-placeholder,      { color: #333 }          /* IE 10+ */
        &:-moz-placeholder,           { color: #333 }          /* Firefox 18- */
    }
    &.darkTheme {
        background-color: #333;
        color: white;
        input[type="text"] {
            color: white;
            &::-webkit-input-placeholder  { color: white }     /* Chrome/Opera/Safari */
            &::-moz-placeholder,          { color: white }     /* Firefox 19+ */
            &:-ms-input-placeholder,      { color: white }     /* IE 10+ */
            &:-moz-placeholder,           { color: white }     /* Firefox 18- */
        }
    }
}
.container.open {
    border-radius: 2px 2px 0 0;
}
.container:hover,
.container.open {
    box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
}

.active {
    outline: 0;
}

.empty {
    outline: 0;
}

.toggleBtn {
    font-size: 20px;
}
