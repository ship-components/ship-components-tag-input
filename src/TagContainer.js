import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Utils } from 'ship-components-utility';
import icon from 'ship-components-icon';

import Dropdown from './Dropdown';
import SelectControls from './Controls';

import css from './TagContainer.css';

/**
 * Filterable Select Dropdown Component
 *
 *   options should be Array<Object> with object fields:
 *   key
 *   title
 *   body
 *   icon (optional node)
 *   optionGroup (optional string)
 *   className (optional string)
 */
export default class TagContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      filterText: '',
      active: false,
      empty: this.isEmpty.call(this),
      dropdownPosTop: '53px'
    };

    this.selectHighlightedItem = this.selectHighlightedItem.bind(this);
    this.highlightPreviousItem = this.highlightPreviousItem.bind(this);
    this.highlightNextItem = this.highlightNextItem.bind(this);
    this.handleDropdownPosition = this.handleDropdownPosition.bind(this);
    this.getVisibleOptions = this.getVisibleOptions.bind(this);
    this.handleOpenDropdown = this.handleOpenDropdown.bind(this);
    this.focusInput = this.focusInput.bind(this);
    this.blurInput = this.blurInput.bind(this);
    this.clearItem = this.clearItem.bind(this);
    this.optionWasSelected = this.optionWasSelected.bind(this);
    this.documentClickHandler = this.documentClickHandler.bind(this);
    this.isEmpty = this.isEmpty.bind(this);
    this.handleToggleDropdown = this.handleToggleDropdown.bind(this);
    this.handleSelectItem = this.handleSelectItem.bind(this);
    this.handleKeyboard = this.handleKeyboard.bind(this);
    this.handleInput = this.handleInput.bind(this);
    this.handleClearItem = this.handleClearItem.bind(this);
    this._toggleSwitch = this._toggleSwitch.bind(this);
  }

  componentDidMount() {
    this.handleDropdownPosition();
    document.addEventListener('click', this.documentClickHandler);

    this._inputField = this.refs.selectControls.refs.filterInput;
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      empty: this.isEmpty(nextProps)
    }, this.handleDropdownPosition);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.documentClickHandler);
  }

  /**
   * Toggle the options dropdown
   * @param  {Event} event
   */
  handleToggleDropdown(event) {
    event.preventDefault();
    if (!this.state.active) {
      this.handleOpenDropdown(event);
    } else {
      this.setState({ active: false });
    }
  }

  /**
   * Opens the options dropdown
   */
  handleOpenDropdown(event) {
    if (this.state.active) {
      return;
    }

    this.setState({ active: true },
      () => {
        // time out because of css transitions.
        setTimeout(this.handleDropdownPosition, this.props.transitionDelay);
        this.focusInput();
      });
    if (typeof this.props.onFocus === 'function') {
      this.props.onFocus(event);
    }
  }

  /**
   * On input, update state: filterText, highlighed item
   * @param  {Event}  event
   */
  handleInput(event) {
    let filterText = event.target.value;

    this.setState({
      filterText: filterText,
      active: true
    }, () => {
      this.handleDropdownPosition();

      if (typeof this.props.onFilter === 'function') {
        this.props.onFilter(filterText, event);
      }
    });
  }

  /**
   * Handler for user choosing an option from dropdown
   * @param  {Item}   option  the target option
   * @param  {Event}  event the click or keyboard event
   */
  handleSelectItem(option, event) {
    event.preventDefault();
    event.stopPropagation();

    if (option === null) {
      return;
    }

    if (typeof this.props.onSelect === 'function') {
      this.props.onSelect(Utils.deepCopy(option));
    }

    this.optionWasSelected();
  }

  /**
   * Peripheral tasks to handleSelectItem
   */
  optionWasSelected() {
    this.setState({
      filterText: this.props.multiple ? '' : this.state.filterText,
      empty: false,
      active: false,
      highlightedOption: null
    }, () => {
      this.handleDropdownPosition();

      if (this.props.multiple) {
        this.focusInput();
      }
    });
  }

  /**
   * Remove an item from the selection
   * @param  {Item}   item  the target item
   * @param  {Event}  event
   */
  handleClearItem(item, event) {
    event.preventDefault();
    event.stopPropagation();

    if (typeof this.props.onDeselect === 'function') {
      this.props.onDeselect(item, this.clearItem);
    } else {
      //note: this will not work properly as is. need to manually update state.selection
      this.clearItem();
    }
  }

  /**
   * Peripheral tasks to handleClearitem
   */
  clearItem() {
    this.setState({
      empty: this.props.selection.length === 0
    }, () => {
      this.handleDropdownPosition();

      if (this.props.multiple) {
        this.focusInput();
      }
    });
  }

  /**
   * Keyboard shortcuts (select item, close dropdown)
   */
  handleKeyboard(event) { // eslint-disable-line complexity
    let code = event.keyCode || event.which;
    switch (code) {
    case 13: // enter
      if (!(this.state.highlightedOption && this.state.active) && typeof this.props.onEnterKey === 'function') {
        this.props.onEnterKey(event);
        break;
      }
    case 9: // tab
      event.preventDefault();
      this.selectHighlightedItem(event);
      break;
    case 38: // up
      event.preventDefault();
      if (this.state.active) {
        this.highlightPreviousItem();
      }
      break;
    case 40: // down
      if (!this.state.active) {
        this.setState({
          active: true
        });
      }
      this.highlightNextItem();
      break;
    case 27: // escape
      this.setState({ active: false }, this.blurInput);
      break;
    }
    event.stopPropagation();
  }

  /**
   * Determines the `top` css property for the dropdown
   */
  handleDropdownPosition() {
    let selectBoxHeight = (this.refs.selectBox.offsetHeight) + 'px';

    // compare strings
    if (selectBoxHeight !== this.state.dropdownPosTop) {
      this.setState({ dropdownPosTop: selectBoxHeight });
    }
  }

  /**
   * Selects the currently highlighted item
   * @param  {Event} event
   */
  selectHighlightedItem(event) {
    let highlightedItem = this.state.highlightedOption;

    if (highlightedItem !== null) {
      this.handleSelectItem(highlightedItem, event);
    }
  }

  /**
   * Highlights one item up in dropdown; meant to be triggered by user pressing `up arrow` key.
   */
  highlightPreviousItem() {
    let highlightedOption = this.state.highlightedOption;

    if (!highlightedOption) {
      return;
    }

    let options = this.getVisibleOptions();
    let currentIndex = options.findIndex(option => option.key === highlightedOption.key);

    if (currentIndex === 0) {
      return;
    }
    highlightedOption = options[currentIndex - 1];
    this.setState({
      highlightedOption: highlightedOption
    });
  }

  /**
   * Highlights one item down in dropdown; meant to be triggered by user pressing `down arrow` key.
   */
  highlightNextItem() {
    let highlightedOption = this.state.highlightedOption;
    let options = this.getVisibleOptions();

    if (!highlightedOption) {
      highlightedOption = options[0];
    } else {
      let currentIndex = options.findIndex(option => option.key === highlightedOption.key);

      if (currentIndex === options.length - 1) {
        return;
      }
      highlightedOption = options[currentIndex + 1];
    }
    this.setState({
      highlightedOption: highlightedOption
    });
  }

  /**
   * Focus cursor on filter input
   */
  focusInput() {
    this.refs.selectControls.focusInput();
  }

  blurInput() {
    this.refs.selectControls.blurInput();
  }

  getFilterResults(filterText) {
    const { options, orderOptionsBy } = this.props;
    const regex = new RegExp(filterText, 'i');

    return options.filter((option) => {
      const orderOptionsByType = typeof option[orderOptionsBy];

      if (orderOptionsByType === 'string') {
        return regex.test(option[orderOptionsBy].toLowerCase());
      }

      if (orderOptionsByType === 'number') {
        return option[orderOptionsBy] === Number(filterText);
      }

      // In case searching through an object
      return option[orderOptionsBy].hasOwnProperty(filterText);
    });
  }

  /**
   * Determines what options to show in dropdown
   * @param  {string}       filterText
   * @return {Array<Option>}  options matching `filterText` and not already selected
   */
  getVisibleOptions(filterText) {
    filterText = typeof filterText === 'undefined' ? this.state.filterText : filterText;
    let options = this.props.options;

    // show filtered results
    if (filterText !== '') {
      options = this.getFilterResults(filterText);
    }

    // exclude selected options
    // TODO : its probably better (in single-select cases) to show the selected item in the dropdown with a checkmark, similar to standard <select />
    if (!this.isEmpty()) {
      if (this.props.selection instanceof Array) {
        if (this.props.selection.length > 0) {
          options = options.filter((item) => {
            let selector = item.key ? 'key' : 'id';
            let index = this.props.selection.findIndex(selectedOption => selectedOption[selector] === item[selector]);
            return index < 0;
          });
        }
      } else {
        options = options.filter(item => item.key !== this.props.selection.key);
      }
    }

    /**
     * Sort by:
     * 1 - "optionGroup"
     * 2 - "score" if possible
     * 3 - props.orderOptionsBy field
     */
    return options.sort((a, b) => { // eslint-disable-line complexity
      if (a.optionGroup !== b.optionGroup) {
        return a.optionGroup > b.optionGroup ? 1 : -1;
      }
      if (a.score && b.score && a.score !== b.score) {
        // lower score = closer match (Fuse.js)
        return a.score > b.score ? 1 : -1;
      }
      if (typeof a[this.props.orderOptionsBy] === 'string') {
        return a[this.props.orderOptionsBy].replace(/\s\s+/g,' ').toLowerCase() > b[this.props.orderOptionsBy].replace(/\s\s+/g,' ').toLowerCase() ? 1 : -1;
      }
      return a[this.props.orderOptionsBy] > b[this.props.orderOptionsBy] ? 1 : -1;
    });
  }

  documentClickHandler(event) {
    if (event.defaultPrevented) {
      return;
    }
    this.setState({
      active: false
    });
  }

  isEmpty(props = this.props) {
    if (props.selection instanceof Array) {
      return props.selection.length === 0;
    } else {
      return !props.selection;
    }
  }

  _toggleSwitch() {
    return <i className={classNames(icon[this.props.toggleSwitchStyle], css.toggleBtn)} />;
  }

  render() {
    const toggleSwitch = this._toggleSwitch();
    let options = this.getVisibleOptions();

    return (
      <div
        ref='selectBox'
        onClick={(event) => {
          event.preventDefault();
          event.stopPropagation();
        }}
        className={classNames(
          {
            [css.active]:this.state.active,
            [css.empty]:this.state.empty,
            [css.darkTheme]: this.props.darkTheme
          },
          css.container,
          this.props.className)}
      >
        <SelectControls
          {...this.props}
          ref='selectControls'
          style={{ top: this.state.dropdownPosTop }}
          isActive={this.state.active}
          isEmpty={this.state.empty}
          filterText={this.state.filterText}
          toggleSwitch={toggleSwitch}
          onToggle={this.handleToggleDropdown}
          onChange={this.handleInput}
          onClear={this.handleClearItem}
          onKeyDown={this.handleKeyboard}
          onFocus={this.handleOpenDropdown}
        />

        <Dropdown
          {...this.props}
          filterText={this.state.filterText}
          open={this.state.active}
          highlightedOption={this.state.highlightedOption}
          options={options}
          onHighlight={this.handleHighlightOption}
          onKeyDown={this.handleKeyboard}
          onSelect={this.handleSelectItem}
        />

      </div>
    );
  }
}

// default props
TagContainer.defaultProps = {
  className:            '',
  transitionDelay:      250,

  onFilter:             function onFilter() {},
  onFocus:              function onFocus() {},
  onEnterKey:           function onEnterKey() {},
  onNewItem:            function onNewItem() {}
};

// prop types checking
TagContainer.propTypes = {
  multiple:           PropTypes.bool.isRequired,
  darkTheme:          PropTypes.bool.isRequired,

  className:          PropTypes.string,
  orderOptionsBy:     PropTypes.string.isRequired,
  placeholder:        PropTypes.string.isRequired,
  togglePosition:     PropTypes.string.isRequired,
  noOptionsMessage:   PropTypes.string.isRequired,
  toggleSwitchStyle:  PropTypes.string.isRequired,

  options:            PropTypes.array.isRequired,
  selection:          PropTypes.array.isRequired,
  optionGroupTitles:  PropTypes.array.isRequired,

  transitionDelay:    PropTypes.number,

  onFilter:           PropTypes.func,
  onFocus:            PropTypes.func,
  onSelect:           PropTypes.func.isRequired,
  onDeselect:         PropTypes.func.isRequired,
  onEnterKey:         PropTypes.func,
  onNewItem:          PropTypes.func
};
