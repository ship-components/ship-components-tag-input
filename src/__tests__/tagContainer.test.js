import React from 'react';
import Immutable from 'immutable';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import TagContainer from '../TagContainer';

const OPTIONS = [
  {
    id: 1,
    title: 'Option 1',
    searchString: 'Option 1'
  },
  {
    id: 2,
    title: 'Option 2',
    searchString: 'Option 2'
  },
  {
    id: 3,
    title: 'Option 3',
    searchString: 'Option 3'
  }
];

// Creates a shallow copy of TagContainer
function initializeShallowTagContainerComponent() {
  const onSelectFn = jest.fn();
  const onDeselectFn = jest.fn();
  const onFetchOptionsFn = jest.fn();

  return {
    wrapper: shallow(
      <TagContainer
        multiple
        filterable
        darkTheme={false}
        orderOptionsBy={'title'}
        label={'label'}
        options={OPTIONS}
        togglePosition={'right'}
        noOptionsMessage={''}
        toggleSwitchStyle={'library_add'}
        selection={new Immutable.List()}
        optionGroupTitles={[]}
        onSelect={onSelectFn}
        onDeselect={onDeselectFn}
        onFetchOptions={onFetchOptionsFn}
      />
    ),
    onSelectFn,
    onDeselectFn,
    onFetchOptionsFn
  };
}

// Creates a mounted copy of TagContainer
function initializeMountTagContainerComponent() {
  const onSelectFn = jest.fn();
  const onDeselectFn = jest.fn();
  const onFetchOptionsFn = jest.fn();

  return {
    wrapper: mount(
      <TagContainer
        multiple
        filterable
        darkTheme={false}
        orderOptionsBy={'title'}
        label={'label'}
        options={OPTIONS}
        togglePosition={'right'}
        noOptionsMessage={''}
        toggleSwitchStyle={'library_add'}
        selection={new Immutable.List()}
        optionGroupTitles={[]}
        onSelect={onSelectFn}
        onDeselect={onDeselectFn}
        onFetchOptions={onFetchOptionsFn}
      />
    ),
    onSelectFn,
    onDeselectFn,
    onFetchOptionsFn
  };
}

describe('Component: TagContainer', () => {
  describe('TagContainer Rendering', () => {
    it('should render correctly', () => {
      const {wrapper} = initializeShallowTagContainerComponent();
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('handleToggleDropdown Function', () => {
    it('should toggle "dropdownOpen" state', () => {
      const {wrapper} = initializeMountTagContainerComponent();
      const event = new Event('look');

      expect(wrapper.state().dropdownOpen).toBeFalsy();
      wrapper.instance().handleToggleDropdown(event);
      expect(wrapper.state().dropdownOpen).toBeTruthy();
    });
  });

  describe('handleInput Function', () => {
    it('should change "filterText" state on onChange event', () => {
      const {wrapper} = initializeMountTagContainerComponent();
      const newFilterText = 'test string';

      expect(wrapper.state().filterText).toEqual('');
      expect(wrapper.state().dropdownOpen).toBeFalsy();

      // Trigger  handleInput handler
      // by firing a change event
      wrapper
        .find('SelectControls')
        .props()
        .onChange({
          target: {
            value: newFilterText
          }
        });

      expect(wrapper.state().filterText).toEqual(newFilterText);
      expect(wrapper.state().dropdownOpen).toBeTruthy();
    });
  });

  describe('handleSelectItem Function', () => {
    it('should pass the selected item to parent on onSelect event', () => {
      const {
        wrapper,
        onSelectFn
      } = initializeMountTagContainerComponent();

      expect(onSelectFn.mock.calls.length).toBe(0);

      wrapper
        .find('Dropdown')
        .props()
        .onSelect(OPTIONS[0], new Event('look'));

      // how to ensure certain function is passed as prop to a component
      expect(wrapper.find('Dropdown')
        .prop('handleSelectItem')).toBe(wrapper.instance().onSelect);

      // Making sure the right option is passed to parent component
      expect(onSelectFn.mock.calls[0][0]).toEqual(OPTIONS[0]);
      expect(onSelectFn.mock.calls.length).toBe(1);
    });

    it('should not call parent onSelect function when option is null', () => {
      const {
        wrapper,
        onSelectFn
      } = initializeMountTagContainerComponent();

      expect(onSelectFn.mock.calls.length).toBe(0);

      wrapper
        .find('Dropdown')
        .props()
        .onSelect(null, new Event('look'));

      // how to ensure certain function is passed as prop to a component
      expect(wrapper.find('Dropdown')
        .prop('handleSelectItem')).toBe(wrapper.instance().onSelect);

      // Since we are not calling the props.onSelect function
      // the function should have no value passed in (empty array)
      expect(onSelectFn.mock.calls).toHaveLength(0);
    });
  });

  describe('optionWasSelected Function', () => {
    it('should update react state when fires', () => {
      const {
        wrapper,
        onSelectFn
      } = initializeMountTagContainerComponent();

      expect(onSelectFn.mock.calls.length).toBe(0);

      wrapper
        .find('Dropdown')
        .props()
        .onSelect(OPTIONS[0], new Event('look'));

      if (wrapper.props().multiple) {
        expect(wrapper.state().filterText).toEqual('');
      } else {
        expect(wrapper.state().filterText).toEqual(wrapper.state().filterText);
      }

      expect(wrapper.state().dropdownOpen).toBeFalsy();
      expect(wrapper.state().dropdownOpen).toBeFalsy();
      expect(wrapper.state().highlightedOption).toBe(null);

      expect(onSelectFn.mock.calls.length).toBe(1);
    });

    it('should call focusInput function when props.multiple is true', () => {
      const { wrapper } = initializeMountTagContainerComponent();

      // mock the focusInput function
      wrapper.instance().focusInput = jest.fn();
      // Update the wrapper to save the mock function
      wrapper.update();

      expect(wrapper.instance().focusInput.mock.calls.length).toBe(0);
      wrapper.instance().optionWasSelected();
      expect(wrapper.instance().focusInput.mock.calls.length).toBe(1);
    });
  });

  describe('handleClearItem Function', () => {
    it('should call the parent onDeselect function to remove the item', () => {
      const {
        wrapper,
        onDeselectFn
      } = initializeMountTagContainerComponent();

      expect(onDeselectFn.mock.calls.length).toBe(0);

      wrapper
        .find('SelectControls')
        .props()
        .onClear(OPTIONS[1], new Event('look'));

      // how to ensure certain function is passed as prop to a component
      expect(wrapper.find('SelectControls')
        .prop('handleClearItem')).toBe(wrapper.instance().onClear);

      // Making sure the right option is passed to parent component
      expect(onDeselectFn.mock.calls[0][0]).toEqual(OPTIONS[1]);
      expect(onDeselectFn.mock.calls.length).toBe(1);
    });
  });

  describe('handleKeyboard Function', () => {
    // This test is commented out for changes in 1.0.4. See README.md.
    // it('should pass the correct item to parent on Tab key press', () => {
    //   const {
    //     wrapper,
    //     onSelectFn
    //   } = initializeMountTagContainerComponent();

    //   expect(onSelectFn.mock.calls.length).toBe(0);

    //   // highlight an option by simulating "DOWN" key, otherwise select function will not be called.
    //   wrapper
    //     .instance()
    //     .handleKeyboard({
    //       keyCode: 40,
    //       preventDefault: jest.fn(),
    //       stopPropagation: jest.fn()
    //     });

    //   wrapper
    //     .instance()
    //     .handleKeyboard({
    //       keyCode: 9,
    //       preventDefault: jest.fn(),
    //       stopPropagation: jest.fn()
    //     });

    //   // how to ensure certain function is passed as prop to a component
    //   expect(wrapper.find('Dropdown')
    //     .prop('handleKeyboard')).toBe(wrapper.instance().onKeyDown);

    //   expect(onSelectFn.mock.calls.length).toBe(1);
    // });

    it('should pass the correct item to parent on Enter key press', () => {
      const {
        wrapper
      } = initializeMountTagContainerComponent();

      wrapper.setProps({
        onEnterKey: jest.fn()
      });

      expect(wrapper.props().onEnterKey.mock.calls.length).toBe(0);

      wrapper
        .instance()
        .handleKeyboard({
          keyCode: 13,
          preventDefault: jest.fn(),
          stopPropagation: jest.fn()
        });

      // how to ensure certain function is passed as prop to a component
      expect(wrapper.find('Dropdown')
        .prop('handleKeyboard')).toBe(wrapper.instance().onKeyDown);

      expect(wrapper.props().onEnterKey.mock.calls.length).toBe(1);
    });

    it('should pass the previous item to parent on Up key press', () => {
      const {
        wrapper
      } = initializeMountTagContainerComponent();

      // Mock highlightPreviousItem function to make sure it's called
      wrapper.instance().highlightPreviousItem = jest.fn();
      expect(wrapper.instance().highlightPreviousItem.mock.calls.length).toBe(0);

      // Setting the dropdownOpen state to true
      // Otherwise highlightPreviousItem will not fires
      wrapper.setState({
        dropdownOpen: true
      });

      wrapper
        .instance()
        .handleKeyboard({
          keyCode: 38,
          preventDefault: jest.fn(),
          stopPropagation: jest.fn()
        });

      // how to ensure certain function is passed as prop to a component
      expect(wrapper.find('Dropdown')
        .prop('handleKeyboard')).toBe(wrapper.instance().onKeyDown);

      expect(wrapper.instance().highlightPreviousItem.mock.calls.length).toBe(1);
    });

    it('should pass the next item to parent on Down key press', () => {
      const {
        wrapper
      } = initializeMountTagContainerComponent();

      // Mock highlightPreviousItem function to make sure it's called
      wrapper.instance().highlightNextItem = jest.fn();
      expect(wrapper.instance().highlightNextItem.mock.calls.length).toBe(0);

      wrapper
        .instance()
        .handleKeyboard({
          keyCode: 40,
          preventDefault: jest.fn(),
          stopPropagation: jest.fn()
        });

      // how to ensure certain function is passed as prop to a component
      expect(wrapper.find('Dropdown')
        .prop('handleKeyboard')).toBe(wrapper.instance().onKeyDown);

      expect(wrapper.instance().highlightNextItem.mock.calls.length).toBe(1);
    });

    it('should change the dropdownOpen state to false on Escape key press', () => {
      const {
        wrapper
      } = initializeMountTagContainerComponent();

      // Mock highlightPreviousItem function to make sure it's called
      wrapper.instance().blurInput = jest.fn();
      expect(wrapper.instance().blurInput.mock.calls.length).toBe(0);

      // Making sure the dropdownOpen state is true
      // Which means the dropdown is open
      wrapper.setState({
        dropdownOpen: true
      });

      wrapper
        .instance()
        .handleKeyboard({
          keyCode: 27,
          preventDefault: jest.fn(),
          stopPropagation: jest.fn()
        });

      // dropdownOpen state should be false when press Escape key
      expect(wrapper.state().dropdownOpen).toBeFalsy();
      expect(wrapper.instance().blurInput.mock.calls.length).toBe(1);
    });
  });

  describe('filter behavior', () => {
    it('does not attempt to fetch options unless filter length > 2', () => {
      const {wrapper, onFetchOptionsFn} = initializeMountTagContainerComponent();
      const shortFilter = '12';
      const adequateFilter = '123';

      wrapper.setProps({autoComplete: true});

      // change filter text
      const selectControls = wrapper
        .find('SelectControls');

      // it shouldn't fetch on filter
      selectControls.props()
        .onChange({
          target: {
            value: shortFilter
          }
        });
      expect(onFetchOptionsFn).not.toHaveBeenCalled();

      // it shouldn't fetch on dropdown opening
      selectControls.props().onFocus();
      expect(onFetchOptionsFn).not.toHaveBeenCalled();

      // it should fetch on filter
      selectControls.props()
        .onChange({
          target: {
            value: adequateFilter
          }
        });

      // it should fetch on dropdown opening
      selectControls.props().onFocus();
      expect(onFetchOptionsFn).toHaveBeenCalled();
    });
  });

  describe('getFilterResults Function', () => {
    it('should return the correct result orderOptionsBy: title', () => {
      const {
        wrapper
      } = initializeMountTagContainerComponent();

      wrapper.setProps({
        options: OPTIONS,
        orderOptionsBy: 'title'
      });

      let getAllResults = wrapper.instance().getFilterResults('option');
      let getOneResult = wrapper.instance().getFilterResults('option 1')[0];

      expect(getAllResults).toEqual(OPTIONS);
      expect(getOneResult).toEqual(OPTIONS[0]);
    });

    it('should return multiple option when orderOptionsBy: title', () => {
      const {
        wrapper
      } = initializeMountTagContainerComponent();

      let sampleOptionList = [
        {
          id: 1,
          title: 'Option 1',
          searchString: 'Option 1'
        },
        {
          id: 10,
          title: 'Option 10',
          searchString: 'Option 10'
        }
      ];

      wrapper.setProps({
        options: sampleOptionList,
        orderOptionsBy: 'title'
      });

      let getTwoResults = wrapper.instance().getFilterResults('1');

      expect(getTwoResults).toEqual(sampleOptionList);
    });

    it('should return the correct result orderOptionsBy: id', () => {
      const {
        wrapper
      } = initializeMountTagContainerComponent();

      wrapper.setProps({
        options: OPTIONS,
        orderOptionsBy: 'id'
      });

      let getOneResult = wrapper.instance().getFilterResults('2')[0];

      expect(getOneResult).toEqual(OPTIONS[1]);
    });
  });

  describe('getVisibleOptions Function', () => {
    it('should return the correct result orderOptionsBy: title', () => {
      const {
        wrapper
      } = initializeMountTagContainerComponent();

      wrapper.setProps({
        options: OPTIONS,
        orderOptionsBy: 'title'
      });

      let getAllResults = wrapper.instance().getVisibleOptions('option');
      let getOneResult = wrapper.instance().getVisibleOptions('option 1')[0];

      expect(getAllResults).toEqual(OPTIONS);
      expect(getOneResult).toEqual(OPTIONS[0]);
    });

    it('should return the correct result sorted', () => {
      const {
        wrapper
      } = initializeMountTagContainerComponent();

      // Create a OrderedSet so we can keep it unsorted
      // Array by default sort the items
      let sampleOptionList = new Immutable.OrderedSet([
        {
          id: 10,
          title: 'Option 10',
          searchString: 'Option 10'
        },
        {
          id: 100,
          title: 'Option 100',
          searchString: 'Option 100'
        },
        {
          id: 1,
          title: 'Option 1',
          searchString: 'Option 1'
        }
      ]);

      wrapper.setProps({
        options: sampleOptionList.toJS(),
        orderOptionsBy: 'title'
      });

      // Getting the sorted results
      let getAllSortedResults = wrapper.instance().getFilterResults('1');

      expect(getAllSortedResults).not.toEqual(sampleOptionList.toJS());
      // Sort sampleOptionList
      var sorted = sampleOptionList.sort((a, b) => a.id > b.id); // eslint-disable-line max-nested-callbacks

      expect(getAllSortedResults).toEqual(sorted.toJS());
    });
  });

  describe('isEmpty Function', () => {
    it('return true when props.selection is Immutable.List && size is zero', () => {
      const { wrapper } = initializeMountTagContainerComponent();
      let sampleProps = {
        selection: new Immutable.List()
      };

      const result1 = wrapper.instance().isEmpty(sampleProps);
      expect(result1).toBeTruthy();

      // Add an item to selection
      // isEmpty should return false
      sampleProps.selection = sampleProps.selection.push(OPTIONS[0]);
      const result2 = wrapper.instance().isEmpty(sampleProps);

      expect(result2).toBeFalsy();
    });

    it('return false when props.selection is not an instance of List', () => {
      const { wrapper } = initializeMountTagContainerComponent();
      let sampleProps = {
        selection: {}
      };

      const result = wrapper.instance().isEmpty(sampleProps);
      expect(result).toBeFalsy();
    });
  });

  describe('highlightPreviousItem Function', () => {
    it('should update state to correct previous item when function calls', () => {
      const { wrapper } = initializeMountTagContainerComponent();

      // setState manually
      wrapper.setState({
        highlightedOption: OPTIONS[2]
      });

      wrapper.instance().highlightPreviousItem();
      expect(wrapper.state().highlightedOption).toEqual(OPTIONS[1]);

    });

    it('should NOT update state to previous item when first item is selected', () => {
      const { wrapper } = initializeMountTagContainerComponent();

      // setState manually
      wrapper.setState({
        highlightedOption: OPTIONS[0]
      });

      wrapper.instance().highlightPreviousItem();
      expect(wrapper.state().highlightedOption).toEqual(OPTIONS[0]);
    });
  });

  describe('highlightNextItem Function', () => {
    it('should update state to correct next item when function calls', () => {
      const { wrapper } = initializeMountTagContainerComponent();

      // setState manually
      wrapper.setState({
        highlightedOption: OPTIONS[0]
      });

      wrapper.instance().highlightNextItem();
      expect(wrapper.state().highlightedOption).toEqual(OPTIONS[1]);

    });

    it('should NOT update state to next item when last item is selected', () => {
      const { wrapper } = initializeMountTagContainerComponent();

      // setState manually
      wrapper.setState({
        highlightedOption: OPTIONS[2]
      });

      wrapper.instance().highlightNextItem();
      expect(wrapper.state().highlightedOption).toEqual(OPTIONS[2]);
    });
  });
});
