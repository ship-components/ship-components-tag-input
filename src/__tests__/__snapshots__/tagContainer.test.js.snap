// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TagContainer TagContainer Rendering should render correctly 1`] = `
<div
  className=""
  onClick={[Function]}
>
  <label
    className=""
  >
    label
  </label>
  <SelectControls
    autoComplete={false}
    className=""
    darkTheme={false}
    filterText=""
    filterable={true}
    isActive={false}
    isEmpty={true}
    label="label"
    loading={false}
    multiple={true}
    noOptionsMessage=""
    onChange={[Function]}
    onClear={[Function]}
    onDeselect={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onSelect={[Function]}
    onToggle={[Function]}
    open={false}
    optionGroupTitles={Array []}
    options={
      Array [
        Object {
          "id": 1,
          "searchString": "Option 1",
          "title": "Option 1",
        },
        Object {
          "id": 2,
          "searchString": "Option 2",
          "title": "Option 2",
        },
        Object {
          "id": 3,
          "searchString": "Option 3",
          "title": "Option 3",
        },
      ]
    }
    orderOptionsBy="title"
    selection={
      Immutable.List [
      ]
    }
    togglePosition="right"
    toggleSwitch={
      <i
        className=""
      />
    }
    toggleSwitchStyle="library_add"
    transitionDelay={250}
  />
  <Dropdown
    autoComplete={false}
    className=""
    darkTheme={false}
    filterText=""
    filterable={true}
    highlightedOption={Object {}}
    label="label"
    loading={false}
    multiple={true}
    noOptionsMessage=""
    onDeselect={[Function]}
    onKeyDown={[Function]}
    onSelect={[Function]}
    open={false}
    optionGroupTitles={Array []}
    options={
      Array [
        Object {
          "id": 1,
          "searchString": "Option 1",
          "title": "Option 1",
        },
        Object {
          "id": 2,
          "searchString": "Option 2",
          "title": "Option 2",
        },
        Object {
          "id": 3,
          "searchString": "Option 3",
          "title": "Option 3",
        },
      ]
    }
    orderOptionsBy="title"
    selection={
      Immutable.List [
      ]
    }
    style={
      Object {
        "top": "53px",
      }
    }
    togglePosition="right"
    toggleSwitchStyle="library_add"
    transitionDelay={250}
  />
</div>
`;
